refactor: Update decoder buffer and logits management

Funatiq opened the pull request:
## Description

- Moved the handling of logits from `DecoderBuffers` to `DecoderInputBuffers`.
- Separated the `DraftBuffers` from `DecoderBuffers` to manage draft token buffers, enhancing the organization of tensor management.
- Removed `DecoderBuffers` from function signatures across various components.
- Updated Python bindings to reflect changes, maintaining compatibility with existing interfaces.

Changes to TRTLLMSampler: 
- Move decoder_state from algs to store.
- Use pure python type for decoder buffer logits.

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

Funatiq committed:
refactor: Update decoder buffer and logits management

- Moved the handling of logits from `DecoderBuffers` to `DecoderInputBuffers`.
- Separated the `DraftBuffers` from `DecoderBuffers` to manage draft token buffers, enhancing the organization of tensor management.
- Removed `DecoderBuffers` from function signatures across various components.
- Updated Python bindings to reflect changes, maintaining compatibility with existing interfaces.

These changes improve the maintainability and clarity of the decoding process in the batch manager.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #5746](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/5746/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #5746](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/5746/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #4201](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/4201/) completed with status: 'FAILURE'

Funatiq commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #5849](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/5849/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #5849](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/5849/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #4286](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/4286/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #6628](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6628/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #6783](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6783/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #6783](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6783/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #4945](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/4945/) completed with status: 'FAILURE'

Funatiq committed:
refactor: Move decoder_state from algs to store

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #7322](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7322/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #7319](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7319/) [ run ] completed with state `ABORTED`

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #7465](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7465/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #5419](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/5419/) completed with status: 'ABORTED'

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #7921](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7921/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #7921](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7921/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #5726](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/5726/) completed with status: 'FAILURE'

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #8332](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8332/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #6035](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/6035/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #8443](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8443/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #8443](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8443/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #6114](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/6114/) completed with status: 'FAILURE'

Funatiq commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #8612](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8612/) [ run ] triggered by Bot

Funatiq committed:
fix: check buffers before access

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run --stage-list "DGX_H100-4_GPUs-PyTorch-DeepSeek-1"

tensorrt-cicd commented:
[PR_Github #8737](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8737/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #8737](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8737/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #6340 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/6340/) completed with status: 'FAILURE'

Funatiq commented:
[PR_Github #8737](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8737/) + [PR_Github #8839](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8839/) passed all tests.

tensorrt-cicd commented:
[PR_Github #8885](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8885/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #8839](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8839/) [ run ] completed with state `ABORTED`

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #8890](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8890/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #8890](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8890/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #6475](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/6475/) completed with status: 'SUCCESS'

Funatiq committed:
refactor: use pure python type for decoder buffer logits

- Modified the `HandleContextLogits` class to accept `max_num_sequences` and return a list of logits tensors along with the logits index.
- Adjusted the `HandleGenerationLogits` class to work with the updated logits handling, removing the dependency on `DecoderInputBuffers`.
- Updated the `TRTLLMSampler` to accommodate changes in logits management, ensuring proper buffer handling.

These changes enhance the clarity and maintainability of the logits processing workflow.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #9058](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/9058/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #6622](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/6622/) completed with status: 'FAILURE'

Funatiq commented:
/bot run

Funatiq commented:
/bot run

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #9231](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/9231/) [ run ] triggered by Bot

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #9235](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/9235/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #6772](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/6772/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/6772/test-results/rerun_report.html) for details.
